client.go

```diff
@@ -1,50 +1,43 @@
 package client
 
 import (
-	"sync"
-
+	"errors"
 	"github.com/hashcloak/Meson/plugin/pkg/common"
-	"github.com/katzenpost/client"
+	katzenClient "github.com/katzenpost/client"
 	"github.com/katzenpost/client/config"
 	"github.com/katzenpost/core/crypto/ecdh"
 	"github.com/katzenpost/core/log"
 	"gopkg.in/op/go-logging.v1"
+	"path/filepath"
+	"sync"
 )
 
 type Client struct {
-	*client.Client
+	*katzenClient.Client
+
 	cfg        *config.Config
 	logBackend *log.Backend
 	log        *logging.Logger
 	fatalErrCh chan error
 	haltedCh   chan interface{}
 	haltOnce   *sync.Once
-	session    *client.Session
+	session    *katzenClient.Session
 	linkKey    *ecdh.PrivateKey
 	service    string
 }
 
-func (c *Client) Start() {
+func (c *Client) Start() error {
+	var err error
 	// Retrieve PKI consensus documents and related info
-	cfg, linkKey := client.AutoRegisterRandomClient(c.cfg)
-	c.cfg = cfg
-	c.linkKey = linkKey
-	session, err := c.NewSession(c.linkKey)
-	if err != nil {
-		panic(err)
-	}
-	c.session = session
-}
-
-func (c *Client) Stop() {
-	// shutdown client and clean up any threads
-	c.Shutdown()
+	_, c.linkKey = katzenClient.AutoRegisterRandomClient(c.cfg)
```

Comment:
`AutoRegisterRandomClient` returns `(config, linkkey)`, where `config` is essentially the same config that you would load in when calling `New` but with extra attributes added to it. So, you can't discard the new `config` that this function returns unless you manually set those attributes in the `config` struct.

Reply:
Yeah, it does return those two values but it also takes in the config as a pointer. This function has a few anti patterns... `AutoRegisterRandomClient` sets the new values of `c.cfg`. So if we do something like this:

```go
func (c *Client) Start() error {
    var err error
    var temp *config.Config
    // Retrieve PKI consensus documents and related info
    temp, c.linkKey = katzenClient.AutoRegisterRandomClient(c.cfg)

    if !reflect.DeepEqual(temp, c.cfg) {
      panic("The return value and config are not the same")
    }

    c.session, err = c.NewSession(c.linkKey)
    return err
}
```

The tests still pass and there is no panic because `AutoRegisterRandomClient` sets the values of `c.cfg` through the pointer. `temp` and `c.fg` reference the same structure.


![tests_002](https://user-images.githubusercontent.com/10123236/70728437-079caf00-1cc7-11ea-8451-e1fdfe127fc9.png)


And here ar the values of the printf:
```
&{Logging:0xc0002de360 UpstreamProxy:0xc0002f2140 Debug:0xc000320000 NonvotingAuthority:0xc0002f46e0 VotingAuthority:<nil> Account:0xc00037c9f0 Registration:0xc000374b60 Panda:0xc0002de3f0 upstreamProxy:0xc0002ee180}

&{Logging:0xc0002de360 UpstreamProxy:0xc0002f2140 Debug:0xc000320000 NonvotingAuthority:0xc0002f46e0 VotingAuthority:<nil> Account:0xc00037c9f0 Registration:0xc000374b60 Panda:0xc0002de3f0 upstreamProxy:0xc0002ee180}
````

Reply:
I understand. I forgot that `AutoRegisterRandomClient` takes in pointers and returns a pointer. So, it's pass by reference and the `config` struct will get modified regardless.